### PR TITLE
Coordinates enhance

### PR DESCRIPTION
**Pull request by kinegratii**
详见 #460 。

补充：

- 没有使用 `CityGeoPoint` 类
- 新模块名称为 `pyecharts.datasets.coordinates`

datasets 包的每个模块均包括两个内容：

- 原始数据
- 访问接口

datasets 将被设计为额外性的组件，功能将从主线 “图表构建 - 环境/引擎加载 - 渲染 ” 脱离。

### CODE REVIEW

File: pyecharts/charts/geolines.py [hunk 1]
```diff
@@ -11,9 +12,31 @@ class GeoLines(Chart):
...
+
+    def get_coordinate(self, name):
+        """
+        Return coordinate for the city name.
+        :param name: City name or any custom name string.
+        :return: A list like [longitude, latitude] or None
+        """
+        if name in self._coordinates:
+            return self._coordinates[name]
+        return get_coordinate(name)
```
Comment (chenjiandongx) [inline pyecharts/charts/geolines.py]: 直接 return None 如何

File: pyecharts/datasets/coordinates.py [hunk 2]
```diff
@@ -0,0 +1,3765 @@
...
+__all__ = ['search_coordinates', 'get_coordinate']
+
+
+def keyword_filter(name, keyword):
+    if not isinstance(keyword, (list, tuple, set)):
+        keyword = keyword,
+    return any([k in name for k in keyword])
+
+
+def search_coordinates(keyword=None, func=None):
```
Comment (chenjiandongx) [inline pyecharts/datasets/coordinates.py]: 这个是否应该在文档注明 两个参数 只取其一

File: pyecharts/datasets/coordinates.py [hunk 3]
```diff
@@ -0,0 +1,3765 @@
...
+The Raw Data and Access Interface for builtin coordinates.
+"""
+from __future__ import unicode_literals
+from functools import partial
+
+__all__ = ['search_coordinates', 'get_coordinate']
+
+
+def keyword_filter(name, keyword):
+    if not isinstance(keyword, (list, tuple, set)):
```
Comment (chenjiandongx) [inline pyecharts/datasets/coordinates.py]: 判断这个的意思是？

File: pyecharts/datasets/coordinates.py (hunk 2 above)
Comment (kinegratii) [reply pyecharts/datasets/coordinates.py]: 文档中已从使用角度详细描述了几种用法，不必显式描述 “两个参数 只取其一” 这几个文字。

File: pyecharts/datasets/coordinates.py [hunk 4]
```diff
@@ -0,0 +1,3765 @@
...
+    """
+    Return coordinate for the city name.
+    :param name: City name or any custom name string.
+    :return: A list like [longitude, latitude] or None
+    """
+    return _COORDINATE_DATASET.get(name, None)
+
+
+_COORDINATE_DATASET = {
+    '阿城': [126.58, 45.32],
```
Comment (chfw) [inline pyecharts/datasets/coordinates.py]: can this be stored in json resource file?

File: pyecharts/datasets/coordinates.py [hunk 5]
```diff
@@ -0,0 +1,3765 @@
...
+"""
+from __future__ import unicode_literals
+from functools import partial
+
+__all__ = ['search_coordinates', 'get_coordinate']
+
+
+def keyword_filter(name, keyword):
+    if not isinstance(keyword, (list, tuple, set)):
+        keyword = keyword,
```
Comment (chfw) [inline pyecharts/datasets/coordinates.py]: extra comma

File: pyecharts/datasets/coordinates.py (hunk 3 above)
Comment (chfw) [reply pyecharts/datasets/coordinates.py]: this function is a private function, so it should start with '_' and be put at the end of this file.

File: pyecharts/datasets/coordinates.py (hunk 4 above)
Comment (chfw) [reply pyecharts/datasets/coordinates.py]: too many records. should not be in a python script file.

Comment (chfw) [reply pyecharts/datasets/coordinates.py]: I knew it was like this previously, but we should build quality into the code step by step.

File: pyecharts/datasets/coordinates.py (hunk 5 above)
Comment (chenjiandongx) [reply pyecharts/datasets/coordinates.py]: 哈哈哈哈 看走眼看走眼

File: pyecharts/datasets/coordinates.py (hunk 4 above)
Comment (chenjiandongx) [reply pyecharts/datasets/coordinates.py]: 现在是从 py 文件直接引入变量使用

而如果是变为 json 文件的话这中间就多了一个步骤了

Comment (chfw) [reply pyecharts/datasets/coordinates.py]: for maintenance, any changes to the json file is not a code change. while with all data inside a python file, any change is a code change.

json file can be validated using external tool easily. but python code changes break all assumptions and need unit tests, integration tests and manual tests.

Comment (chfw) [reply pyecharts/datasets/coordinates.py]: it is more expensive to tests if we keep the data inside python file than keeping them in a json file.

**Comment by kinegratii** [comment]
`search_coordinates` function as the following:
https://github.com/pyecharts/pyecharts/blob/2243a89761f3862276960cfeb0d4aeb0d284b8c7/pyecharts/datasets/coordinates.py#L10-L26

1 Move L18-L19 to before the filter function define,And then these code will be run only once.
2 Now function `keyword_filter` has only one line code, it is better to move inside `search_coordinates` and `functools.partial` can be also removed.

**Comment by kinegratii** [comment]
## About Store of Coordinate Data
(1) I think it is very correct to make **coordinate data** store in external data files, as well as **map geojson data**.
(2) It is better to use **lazy-load:When need,when load ** to load these external data,because these data will be used in a small amount of chart render (2/28).
(3) So It need another PR to improve the lazy load.

**Comment by chenjiandongx** [comment]
@chfw 你还有什么意见吗？

File: pyecharts/datasets/coordinates.py [hunk 6]
```diff
@@ -0,0 +1,3762 @@
+# coding=utf8
+"""
+The Raw Data and Access Interface for builtin coordinates.
+"""
+from __future__ import unicode_literals
+
+__all__ = ['search_coordinates', 'get_coordinate']
+
+
+def search_coordinates(keyword=None, func=None):
```
Comment (chfw) [inline pyecharts/datasets/coordinates.py]: 这个 search_coordinates 有没有更简单的实现方式？

Comment (chfw) [reply pyecharts/datasets/coordinates.py]: 而且 search_coordinates() 会出现问题

**Comment by chfw** [comment]
search_coordinates 需要再磨一磨

File: pyecharts/datasets/coordinates.py [hunk 7]
```diff
@@ -0,0 +1,3770 @@
...
+from __future__ import unicode_literals
+
+__all__ = [
+    'search_coordinates_by_filter',
+    'search_coordinates_by_keyword',
+    'get_coordinate'
+]
+
+
+def search_coordinates_by_filter(func):
```
Comment (chfw) [inline pyecharts/datasets/coordinates.py]: Explicit is better than implicit.

File: docs/zh-cn/datasets.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+### 按过滤函数搜索地理坐标
+
+`search_coordinates_by_filter(func)` 根据过滤函数，返回一个匹配的字典对象。
+用法（结果同上）
+
+```python
+from pyecharts.datasets.coordinates import search_coordinates_by_filter
+
+result = search_coordinates_by_filter(
+    func=lambda name: '福州' in name or '杭州' in name
```
Comment (chfw) [inline docs/zh-cn/datasets.md]: func is not compulsory, but looks good.

**Pull request merged**